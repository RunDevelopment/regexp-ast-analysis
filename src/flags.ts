/**
 * A simple interface to represent JS RegExp flags.
 *
 * All properties are optional and assumed to be `false` by default.
 */
export interface ReadonlyFlags {
	/**
	 * The `s` flag.
	 *
	 * @default false
	 */
	readonly dotAll?: boolean;
	/**
	 * The `g` flag.
	 *
	 * @default false
	 */
	readonly global?: boolean;
	/**
	 * The `d` flag.
	 *
	 * @default false
	 */
	readonly hasIndices?: boolean;
	/**
	 * The `i` flag.
	 *
	 * @default false
	 */
	readonly ignoreCase?: boolean;
	/**
	 * The `m` flag.
	 *
	 * @default false
	 */
	readonly multiline?: boolean;
	/**
	 * The `y` flag.
	 *
	 * @default false
	 */
	readonly sticky?: boolean;
	/**
	 * The `u` flag.
	 *
	 * @default false
	 */
	readonly unicode?: boolean;
}
