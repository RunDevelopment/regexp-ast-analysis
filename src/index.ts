export * from "./basic";
export * from "./to-char-set";
export * from "./chars";
export * from "./flags";
export * from "./follow";
export * from "./next-char";
