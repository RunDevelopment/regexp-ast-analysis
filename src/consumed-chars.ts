import { Alternative, Element, Pattern } from "@eslint-community/regexpp/ast";
import { CharSet } from "refa";
import { ReadonlyFlags } from "./flags";
import { hasSomeDescendant, isEmptyBackreference } from "./basic";
import { Chars } from "./chars";
import { toUnicodeSet } from "./to-char-set";

export interface ConsumedChars {
	chars: CharSet;
	/**
	 * Whether `char` is exact.
	 *
	 * If `false`, then `char` is only guaranteed to be a superset of the
	 * actually possible characters.
	 */
	exact: boolean;
}

/**
 * Returns the union of all characters that can possibly be consumed by the
 * given element.
 */
export function getConsumedChars(element: Element | Pattern | Alternative, flags: ReadonlyFlags): ConsumedChars {
	const sets: CharSet[] = [];
	let exact = true;

	// we misuse hasSomeDescendant to iterate all relevant elements
	hasSomeDescendant(
		element,
		d => {
			if (
				d.type === "Character" ||
				d.type === "CharacterClass" ||
				d.type === "CharacterSet" ||
				d.type === "ExpressionCharacterClass"
			) {
				const c = toUnicodeSet(d, flags);

				sets.push(c.chars);
				if (!c.accept.isEmpty) {
					const chars = new Set<CharSet>();
					for (const word of c.accept.wordSets) {
						for (const char of word) {
							chars.add(char);
						}
					}
					sets.push(Chars.empty(flags).union(...chars));
				}

				exact = exact && !c.isEmpty;
			} else if (d.type === "Backreference" && !isEmptyBackreference(d, flags)) {
				const c = getConsumedChars(d.resolved, flags);
				sets.push(c.chars);
				exact = exact && c.exact && c.chars.size < 2;
			}

			// always continue to the next element
			return false;
		},
		// don't go into assertions
		d => {
			if (d.type === "CharacterClass" || d.type === "ExpressionCharacterClass") {
				return false;
			}
			if (d.type === "Assertion") {
				exact = false;
				return false;
			}
			return true;
		}
	);

	const chars = Chars.empty(flags).union(...sets);

	return { chars, exact };
}
