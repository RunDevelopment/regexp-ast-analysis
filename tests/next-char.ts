import { assert } from "chai";
import { RegExpParser, visitRegExpAST } from "regexpp";
import * as RAA from "../src";
import { MatchingDirection } from "../src";
import { selectNamedGroups } from "./helper/select";
import { assertSnapshot } from "./helper/snapshot";

function* iter<T>(array: T | T[]): IterableIterator<T> {
	if (Array.isArray(array)) {
		yield* array;
	} else {
		yield array;
	}
}

describe(RAA.getFirstConsumedChar.name, function () {
	interface TestCase {
		regexp: RegExp | RegExp[];
		direction?: MatchingDirection;
	}

	test([
		{ regexp: /a/ },
		{ regexp: /a/i },
		{ regexp: /[a-z]/i },
		{ regexp: /[\0-\uFFFF]/ },
		{ regexp: /[\0-\u{10FFFF}]/u },

		{ regexp: /abc/ },
		{ regexp: /abc/, direction: "ltr" },
		{ regexp: /abc/, direction: "rtl" },

		{ regexp: [/a?/, /a??/, /a|/, /|a/, /a*/, /a+|(|)/, /(?:a+){0,4}/, /a?b{0}/] },

		{ regexp: /(?:)/ },

		// backreferences

		{ regexp: /(a)\1/ },
		{ regexp: /(a)\1/, direction: "rtl" },
		{ regexp: /(?:(a)|b)\1/, direction: "rtl" },
		{ regexp: /(a)b\1/ },
		{ regexp: /\1(a)/ },
		{ regexp: /\1a|a(b)/ },

		// assertions

		{ regexp: /^/ },
		{ regexp: /^/, direction: "rtl" },
		{ regexp: /$/ },
		{ regexp: /$/, direction: "rtl" },

		{ regexp: /^/m },
		{ regexp: /$/m },

		{ regexp: /(?=a)/m },
		{ regexp: /(?!a)/m },
		{ regexp: /(?=abc)/m },
		{ regexp: /(?!abc)/m },

		{ regexp: /(?=$)/m },
		{ regexp: /(?!$)/m },

		{ regexp: /(?!a)[ab]/ },
		{ regexp: /(?![])[ab]/ },
		{ regexp: /(?!a?b)[ab]/ },
		{ regexp: /(?!abba)[ab]/ },
		{ regexp: /(?=a)[ab]/ },
		{ regexp: /(?=a?)[ab]/ },
		{ regexp: /(?<!a)[ab]/ },
		{ regexp: /(?<=a)[ab]/ },
		{ regexp: /\b[ab]/ },

		{ regexp: /\b/ },
		{ regexp: /\B/ },
		{ regexp: /^\b/ },
		{ regexp: /^\B/ },

		// make exact again
		{ regexp: /\b[ab]|[ab]/ },
		{ regexp: /\b[ab]|a|b/ },

		// trivially rejecting
		{ regexp: /(?!a?)[ab]/ },

		{ regexp: /(?!a)|b|(?=n)/ },
		{ regexp: /(?<!a)|b|(?=n)/ },

		// This is interesting because the `(?=b)` and `(?=c)` contradict which causes `a` to be the first char
		{ regexp: /(?:a|(?=b))(?=c)/ },
		{ regexp: /(?:a|(?=b))(?!b)/ },
	]);

	function test(cases: TestCase[]): void {
		for (const { regexp, direction = "ltr" } of cases) {
			for (const r of iter(regexp)) {
				it(`${r} (${direction})`, function () {
					const { pattern, flags } = new RegExpParser().parseLiteral(r.toString());

					assertSnapshot(RAA.getFirstConsumedChar(pattern.alternatives, direction, flags));
					// assert.deepEqual(RAA.getFirstConsumedChar(pattern.alternatives, direction, flags), expected);
				});
			}
		}
	}

	it("performance test", function () {
		this.timeout(1000);

		const monster = /(:\s*)(?!\s)(?:!?\s*(?:(?:\?|\bp>|(?:\[\]|\*(?!\*)|\*\*)(?:\s*a\)|\s*ct\b|\s*ve\b|\s*ao\b)*)\s*)*(?:\bpe\b|(?:\be\.)?\b(?!\bk\b)(?!\d)\w+\b(?:\.\b(?!\bk\b)(?!\d)\w+\b)*(?!\s+\b(?!\bk\b)(?!\d)\w+\b)))+(?=\s*(?:a\)\s*)?[=;,)])|(?!\s)(?:!?\s*(?:(?:\?|\bp>|(?:\[\]|\*(?!\*)|\*\*)(?:\s*a\)|\s*ct\b|\s*ve\b|\s*ao\b)*)\s*)*(?:\bpe\b|(?:\be\.)?\b(?!\bk\b)(?!\d)\w+\b(?:\.\b(?!\bk\b)(?!\d)\w+\b)*(?!\s+\b(?!\bk\b)(?!\d)\w+\b)))+(?=\s*(?:a\)\s*)?\{)/;

		const { pattern, flags } = new RegExpParser().parseLiteral(monster.toString());

		type WordAssertionResult = ["all" | "word", "all" | "word"];
		const expected: Record<number, WordAssertionResult> = {
			"30": ["all", "word"],
			"72": ["word", "all"],
			"80": ["word", "all"],
			"88": ["word", "all"],
			"101": ["all", "word"],
			"105": ["word", "all"],
			"111": ["all", "word"],
			"118": ["all", "word"],
			"123": ["word", "word"],
			"126": ["word", "all"],
			"138": ["word", "all"],
			"145": ["word", "word"],
			"150": ["word", "word"],
			"153": ["word", "all"],
			"165": ["word", "all"],
			"175": ["word", "word"],
			"180": ["word", "word"],
			"183": ["word", "all"],
			"195": ["word", "all"],
			"249": ["all", "word"],
			"291": ["word", "all"],
			"299": ["word", "all"],
			"307": ["word", "all"],
			"320": ["all", "word"],
			"324": ["word", "all"],
			"330": ["all", "word"],
			"337": ["all", "word"],
			"342": ["word", "word"],
			"345": ["word", "all"],
			"357": ["word", "all"],
			"364": ["word", "word"],
			"369": ["word", "word"],
			"372": ["word", "all"],
			"384": ["word", "all"],
			"394": ["word", "word"],
			"399": ["word", "word"],
			"402": ["word", "all"],
			"414": ["word", "all"],
		};
		const actual: Record<number, WordAssertionResult> = {};

		visitRegExpAST(pattern, {
			onAssertionEnter(node) {
				if (node.kind !== "word") return;

				const ltr = RAA.getFirstConsumedChar(node, "ltr", flags);
				const rtl = RAA.getFirstConsumedChar(node, "rtl", flags);
				if (!ltr.empty || !rtl.empty) {
					assert.fail("What?");
				}

				actual[node.start] = [ltr.look.char.isAll ? "all" : "word", rtl.look.char.isAll ? "all" : "word"];
			},
		});

		assert.deepStrictEqual(actual, expected);
	});
});

describe(RAA.getFirstConsumedCharAfter.name, function () {
	interface TestCase {
		regexp: RegExp | RegExp[];
		direction?: MatchingDirection;
	}

	test([
		{ regexp: /(?<afterThis>)a/ },
		{ regexp: /(?<afterThis>)a/i },

		{ regexp: /a(?<afterThis>)bc/ },
		{ regexp: /a(?<afterThis>)bc/, direction: "ltr" },
		{ regexp: /a(?<afterThis>)bc/, direction: "rtl" },

		{ regexp: /(?<afterThis>)(?:a?|b)cd/ },

		{ regexp: /(abc(?<afterThis>)){0,1}e/ },
		{ regexp: /(abc(?<afterThis>)){0,2}e/ },
		{ regexp: /(?:a(?:b(?<afterThis>)){2}){2}z/ },

		{
			// This is interesting because `followPaths` used to be implemented in a way that caused this pattern to
			// create exponentially many paths, taking exponentially much time.
			regexp: /(a(b(c(d(e(f(g(h(i(j(k(l(m(n(o(p(q(r(s(t(u(v(w(x(y(?<afterThis>)){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}z/,
		},

		{ regexp: /(?<afterThis>)/ },
		{ regexp: /(?=(?<afterThis>))/ },

		// word boundary assertions
		{ regexp: /(?<afterThis>a|b)\b/ },
		{ regexp: /(?:a|b)(?<afterThis>)\b/ },
		{ regexp: /(?:a|b(?<afterThis>))\b/ },
		{ regexp: /;(?<afterThis>)\b[\d.]+/ },
		{ regexp: /;(?<afterThis>)\b\b\b\b\b\b[\d.]+/ },
		{ regexp: /;(?<afterThis>)\B[\d.]+/ },
		{ regexp: /;(?<afterThis>)\B\B\B[\d.]+/ },
		{ regexp: /^(?<afterThis>)\b/ },
		{ regexp: /a(?=(?<afterThis>)b)/, direction: "ltr" },
		{ regexp: /a(?=(?<afterThis>)b)/, direction: "rtl" },

		{ regexp: /(a)(?<afterThis>)\1/ },
		{ regexp: /(a|b)(?<afterThis>)\1/ },
		{ regexp: /(a|b)|(?<afterThis>)\1/ },
	]);

	function test(cases: TestCase[]): void {
		for (const { regexp, direction = "ltr" } of cases) {
			for (const r of iter(regexp)) {
				it(`${r} (${direction})`, function () {
					const { pattern, flags } = new RegExpParser().parseLiteral(r.toString());

					const [marker] = selectNamedGroups(pattern, /^afterThis$/);

					const actual = RAA.getFirstConsumedCharAfter(marker, direction, flags);
					assertSnapshot(actual);

					assert.deepEqual(
						RAA.getFirstConsumedCharAfterWithContributors(marker, direction, flags).char,
						actual
					);
				});
			}
		}
	}
});

describe(RAA.getFirstCharAfter.name, function () {
	interface TestCase {
		regexp: RegExp | RegExp[];
		direction?: MatchingDirection;
	}

	test([
		{ regexp: /(?<afterThis>)a/ },
		{ regexp: /(?<afterThis>)a/i },

		// The implementation of `getFirstCharAfter` is extremely boring.
		// Its correctness is already ensured by the tests above.
	]);

	function test(cases: TestCase[]): void {
		for (const { regexp, direction = "ltr" } of cases) {
			for (const r of iter(regexp)) {
				it(`${r} (${direction})`, function () {
					const { pattern, flags } = new RegExpParser().parseLiteral(r.toString());

					const [marker] = selectNamedGroups(pattern, /^afterThis$/);

					const actual = RAA.getFirstCharAfter(marker, direction, flags);
					assertSnapshot(actual);
					assert.deepEqual(RAA.getFirstCharAfterWithContributors(marker, direction, flags).char, actual);
				});
			}
		}
	}
});
