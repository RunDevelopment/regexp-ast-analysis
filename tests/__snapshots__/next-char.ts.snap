// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getFirstCharAfter /(?<afterThis>)a/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "edge": false,
  "exact": true,
}
`;

exports[`getFirstCharAfter /(?<afterThis>)a/i (ltr) 1`] = `
Object {
  "char": CharSet (65535) [41, 61],
  "edge": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /$/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /$/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /$/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [a, d, 2028..2029],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?!$)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /(?![])[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?!a)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..60, 62..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?!a)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?!a)|b|(?=n)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..60, 62..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?!a?)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?!a?b)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedChar /(?!abba)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedChar /(?!abc)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /(?:(a)|b)\\1/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?:)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?:a+){0,4}/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?:a|(?=b))(?!b)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?:a|(?=b))(?=c)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?<!a)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedChar /(?<!a)|b|(?=n)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /(?<=a)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedChar /(?=$)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [a, d, 2028..2029],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?=a)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [61],
    "edge": false,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(?=a)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?=a?)[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(?=abc)/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [61],
    "edge": false,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /(a)\\1/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(a)\\1/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /(a)b\\1/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /[\\0-\\u{10FFFF}]/u (ltr) 1`] = `
Object {
  "char": CharSet (1114111) [0..10ffff],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /[\\0-\\uFFFF]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [0..ffff],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /[a-z]/i (ltr) 1`] = `
Object {
  "char": CharSet (65535) [41..5a, 61..7a],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /\\1(a)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /\\1a|a(b)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /\\B/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /\\b[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedChar /\\b[ab]|[ab]/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /\\b[ab]|a|b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /^/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /^/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /^/m (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /^\\B/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..2f, 3a..40, 5b..5e, 60, 7b..ffff],
    "edge": true,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /^\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [30..39, 41..5a, 5f, 61..7a],
    "edge": false,
    "exact": false,
  },
}
`;

exports[`getFirstConsumedChar /|a/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a*/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a+|(|)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /a/i (ltr) 1`] = `
Object {
  "char": CharSet (65535) [41, 61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /a?/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a??/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a?b{0}/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /a|/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedChar /abc/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /abc/ (ltr) 2`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedChar /abc/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [63],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(?:a(?:b(?<afterThis>)){2}){2}z/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62, 7a],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(?:a|b(?<afterThis>))\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..2f, 3a..40, 5b..5e, 60, 7b..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(?:a|b)(?<afterThis>)\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..2f, 3a..40, 5b..5e, 60, 7b..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(?<afterThis>)(?:a?|b)cd/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..63],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(?<afterThis>)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(?<afterThis>)a/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(?<afterThis>)a/i (ltr) 1`] = `
Object {
  "char": CharSet (65535) [41, 61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(?<afterThis>a|b)\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..2f, 3a..40, 5b..5e, 60, 7b..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(?=(?<afterThis>))/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(a(b(c(d(e(f(g(h(i(j(k(l(m(n(o(p(q(r(s(t(u(v(w(x(y(?<afterThis>)){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}){2}z/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..7a],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(a)(?<afterThis>)\\1/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(a|b)(?<afterThis>)\\1/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61..62],
  "empty": false,
  "exact": false,
}
`;

exports[`getFirstConsumedCharAfter /(a|b)|(?<afterThis>)\\1/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [0..ffff],
    "edge": true,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /(abc(?<afterThis>)){0,1}e/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [65],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /(abc(?<afterThis>)){0,2}e/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [61, 65],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /;(?<afterThis>)\\B[\\d.]+/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [2e],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /;(?<afterThis>)\\B\\B\\B[\\d.]+/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [2e],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /;(?<afterThis>)\\b[\\d.]+/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [30..39],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /;(?<afterThis>)\\b\\b\\b\\b\\b\\b[\\d.]+/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [30..39],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /^(?<afterThis>)\\b/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [],
  "empty": true,
  "exact": true,
  "look": Object {
    "char": CharSet (65535) [30..39, 41..5a, 5f, 61..7a],
    "edge": false,
    "exact": true,
  },
}
`;

exports[`getFirstConsumedCharAfter /a(?<afterThis>)bc/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /a(?<afterThis>)bc/ (ltr) 2`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /a(?<afterThis>)bc/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /a(?=(?<afterThis>)b)/ (ltr) 1`] = `
Object {
  "char": CharSet (65535) [62],
  "empty": false,
  "exact": true,
}
`;

exports[`getFirstConsumedCharAfter /a(?=(?<afterThis>)b)/ (rtl) 1`] = `
Object {
  "char": CharSet (65535) [61],
  "empty": false,
  "exact": true,
}
`;
